src/budget/budgetmanager.cpp

```diff
@@ -1059,6 +1081,30 @@ int CBudgetManager::ProcessFinalizedBudgetVote(CFinalizedBudgetVote& vote, CNode
     }
 
     const CTxIn& voteVin = vote.GetVin();
+
+    // See if this vote was signed with a deterministic masternode
+    auto mnList = deterministicMNManager->GetListAtChainTip();
+    auto dmn = mnList.GetMNByCollateral(voteVin.prevout);
+    if (dmn) {
+        if (!vote.CheckSignature(dmn->pdmnState->keyIDOperator)) {
+            LogPrintf("fbvote - signature invalid from dmn %s\n", dmn->proTxHash.ToString());
+            return 100;
+        }
+        AddSeenFinalizedBudgetVote(vote);
+        std::string strError = "masternode not valid or PoSe banned";
+        if (mnList.IsMNValid(dmn) && UpdateFinalizedBudget(vote, pfrom, strError)) {
```

Comment:
Same as with the votes, better to raise the DoS level if the DMN is invalid and add the budget finalization to the seen map after the DMN validity check.